OrbitVulkanLayerClientGgp/OrbitCaptureClientLayer.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <assert.h>
+#include <string.h>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitVulkanLayerClientGgp/DispatchTable.h"
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "vulkan/vk_layer.h"
+#include "vulkan/vk_layer_dispatch_table.h"
+#include "vulkan/vulkan.h"
+
+#undef VK_LAYER_EXPORT
+#if defined(WIN32)
+#define VK_LAYER_EXPORT extern "C" __declspec(dllexport)
+#else
+#define VK_LAYER_EXPORT extern "C"
+#endif
+
+// Layer information
+static constexpr char const* kLayerName = "VK_LAYER_ORBIT_CAPTURE_CLIENT";
+static constexpr char const* kLayerDescription =
+    "Layer that contains Orbit Client implementation to run captures";
+static constexpr uint32_t const kLayerImplementationVersion = 1;
+static constexpr uint32_t const kLayerSpecVersion = VK_API_VERSION_1_1;
+
+// Vulkan layer
+absl::Mutex layer_mutex;
```

Comment:
I considered moving it inside the class  when I cleaned the code but I realised the mutex was needed in the main file as well to read and process the data (e.g command buffer) so I decided to leave it in the main class as the same lock is used in both cases. I considered passing it as a parameter for example and lock inside the DispatchTable methods but I read that was not recommended so that's why I block it before the call.

Reply:
I can update the code but I don't understand how I could lock the data safely in the main file if the mutex are different inside DispatchTable and there; they would not be locked safely in that case, would they?

Reply:
Couldn't you use a different mutex for processing the data? Having one mutex for different data might only be relevant if there might be race conditions between the data.

Reply:
But as I said, it depends a bit of the size of your final layer. If it stays relatively small, it might be totally fine to have one mutex in the main.

Reply:
I checked why in the tutorial they use a global mutex for both the table and the data and the reason they give is "We'll protect this by the same global lock as before to ensure multithreaded recording doesn't cause problems. " so I think it is needed to use the same lock.

Reply:
But you don't have any data to record so far, right? Anyways. For now it is also fine for me. Just wanted to check.